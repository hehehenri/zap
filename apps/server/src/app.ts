import Koa from "koa";
import http from "http";
import cors from "koa-cors";
import bunyan from "bunyan";
import koaBunyan from "koa-bunyan-logger";
import routes from "./routes";
import { KoaContext } from "./schemas/context";
import { execute, subscribe } from "graphql";
import { WebSocketServer } from "ws";
import schema from "./schemas";
import { errorHandlerMiddleware } from "./routes/middlewares";
import { useServer } from "graphql-ws/lib/use/ws";
import config from "./config";

const logger = () => {  
  const options = {
    name: 'zap',
    serializers: bunyan.stdSerializers
  };
  
  const logger = bunyan.createLogger(options);
  return koaBunyan(logger);
}

const createApp = () => {
  const app = new Koa<Koa.DefaultState, KoaContext>();

  app.use(cors({ credentials: true }));
  app.use(logger());

  const router = routes();  
  app.use(router.routes());
  app.use(errorHandlerMiddleware());

  return app;
}

const start = () => {
  const app = createApp();
  const server = http.createServer(app.callback());
  
  server.listen(config.app.port, () => {
    console.log("http: server running on: http://localhost:8000/graphql");

    const wsServer = new WebSocketServer({
      server,
      path: "/graphql"
    });

    useServer({ schema, execute, subscribe }, wsServer)
    console.log("ws: server running on ws://localhost:8000/graphql")
  });
};

export { start }
