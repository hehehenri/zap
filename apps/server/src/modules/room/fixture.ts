import { RoomDefinition, RoomModel } from "./RoomModel";

type Fields = Pick<RoomDefinition, Exclude<keyof RoomDefinition, "_id">>

export const createRoom = async (fields?: DeepPartial<Fields>) => {
  return await new RoomModel(fields).save();
}
