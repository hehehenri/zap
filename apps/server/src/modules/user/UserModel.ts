import mongoose from "mongoose";

export type UserDefinition = {
  readonly _id: mongoose.Types.ObjectId;
  readonly username: string
  readonly password: string
}

export const userSchema = new mongoose.Schema<UserDefinition>(
  {
    username: { type: String, required: true },
    password: { type: String, required: true, select: false },
  },
  { collection: 'User' }
);

export const UserModel = mongoose.model<UserDefinition>('User', userSchema);
