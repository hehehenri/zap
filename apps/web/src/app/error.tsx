"use client";

import Link from "next/link";
import { useRouter } from "next/navigation";

export default function Error({ error }: { error: Error }) {
  const router = useRouter();

  if (error.cause instanceof Response && error.cause.status === 401) {
    router.push("/login");
  }

  return (
    <section className="w-full h-full flex items-center justify-center flex-col gap-y-0.5">
      <h2 className="text-secondary-400 text-2xl lg:text-4xl font-mono">
        Something went wrong!
      </h2>
      <span className="flex gap-1 items-center text-stone-700 text-sm lg:text-lg">
        This is still a working in progress. You can
        <Link
          href="https://twitter.com/hnrbs_"
          className="text-secondary-600 hover:text-secondary-500 hover:-translate-y-0.5 transition"
        >
          message me
        </Link>
        or
        <Link
          href="https://github.com/hnrbs/zap/issues"
          className="text-secondary-600 hover:text-secondary-500 hover:-translate-y-0.5 transition"
        >
          open an issue.
        </Link>
      </span>
    </section>
  );
}
