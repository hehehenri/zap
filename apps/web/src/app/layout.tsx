import "@/styles/globals.css";
import type { Metadata } from "next";
import localFont from "next/font/local";

const bevellier = localFont({
  src: "../../public/bevellier.ttf",
  variable: "--font-bevellier",
});

const satoshi = localFont({
  src: "../../public/satoshi.ttf",
  variable: "--font-satoshi",
});

export const metadata: Metadata = {
  title: "Create Turborepo",
  description: "Generated by create turbo",
};

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}): JSX.Element {
  return (
    <html lang="en" className="bg-secondary-50/70">
      <body className={`${bevellier.variable} ${satoshi.variable} font-sans`}>
        {children}
      </body>
    </html>
  );
}
