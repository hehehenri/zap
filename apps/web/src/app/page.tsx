import { ArrowRight } from "lucide-react";
import Image from "next/image";

const Home = () => {
  return (
    <div className="px-12 py-4 flex flex-col gap-8">
      <header className="inset-x-0 top-0 z-10">
        <nav className="flex items-center justify-between">
          <span className="font-mono text-4xl font-bold text-primary-500">
            Zap
          </span>
          <a
            href="#"
            className="
              flex items-center gap-2
              bg-primary-500 rounded-full py-3 px-7 border border-black
            "
          >
            Sign Up
            <ArrowRight strokeWidth={1} />
          </a>
        </nav>
      </header>

      <div className="pt-32 pb-14 sm:py-48 bg-cyan-950/50 rounded-3xl relative overflow-hidden text-white">
        <div className="mx-auto max-w-6xl px-12 grid xl:grid-cols-2">
          <div>
            <h1 className="text-4xl sm:text-5xl md:text-7xl font-sans font-semibold tracking-tight text-zinc-50 leading-tight">
              Send messages (not) privately
            </h1>
            <p className="mt-6 text-lg sm:text-xl leading-8 tracking-tighter">
              Simple, reliable, private messaging and calling for free*,
              available all over the world.
            </p>

            <a
              href="#"
              className="
                flex items-center justify-center gap-2 w-40 text-black
                bg-primary-500 py-3 px-7 mt-24 rounded-full border border-black
              "
            >
              Sign Up
              <ArrowRight strokeWidth={1} />
            </a>
          </div>
        </div>

        <Image
          layout="fill"
          objectFit="cover"
          alt="mountains"
          src={"/trees.jpg"}
          className="absolute -z-10"
        />
      </div>
    </div>
  );
};

export default Home;
