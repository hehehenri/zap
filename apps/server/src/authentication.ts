import jwt from "jsonwebtoken";
import { UserModel, UserDocument } from "./modules/user/UserModel";
import config from "./config";
import { KoaContext } from "./context";

export type Config = {
  secret: string
}

type TokenPayload = {
  id: string
}

export const getToken = (ctx: KoaContext) => {
  const auth = ctx.headers.authorization;
  if (!auth) return null;

  if (auth.startsWith("Bearer ")) {
    return auth.substring(7, auth.length);
  }

  return null
}

export const getAuth = async (token: string | null) => {
  if (!token) return { user: null };

  try {
    const decoded = jwt.verify(token, config.jwt.secret) as TokenPayload;

    const user = await UserModel.findById(decoded.id);
    return { user };
  } catch (_) {
    return { user: null };
  }
}

export const generateToken = (user: UserDocument) => {
  return jwt.sign({ id: user._id }, config.jwt.secret);
}
