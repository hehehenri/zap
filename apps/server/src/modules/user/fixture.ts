import config from "../../config";
import { UserDefinition, UserModel } from "./UserModel"
import bcrypt from "bcryptjs";

let counter = 0;

const getCounter = () => {
  counter += 1;

  return counter;
}

type Fields = Pick<UserDefinition, Exclude<keyof UserDefinition, "_id">>

export const createUser = async (fields?: DeepPartial<Fields>) => {
  const index = getCounter();

  const plainTextPassword = fields?.password ?? "password";

  const password = bcrypt.hashSync(plainTextPassword, config.jwt.saltRounds);
  const username = fields?.username ?? `user#${index}`;

  const user = new UserModel({ password, username });
  await user.save();

  return user;
}
