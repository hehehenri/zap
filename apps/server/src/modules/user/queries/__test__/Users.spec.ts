import { UserDefinition } from "../../UserModel";
import { generateToken } from "../../../../authentication";
import request from "supertest";
import { createApp } from "../../../../app";
import { createUser } from "../../fixture";
import { database } from "../../../../test";
import { MongoMemoryServer } from "mongodb-memory-server";

describe('user/queries/users', () => {
  let db: MongoMemoryServer;
  
  beforeAll(async () => {
    db = await MongoMemoryServer.create();
    database.connect(db) 
  });
  afterAll(async () => database.disconnect(db));
  afterEach(database.clear);
  
  it("should return a collection of all users", async () => {
    const firstUser: UserDefinition = await createUser(); 
    await createUser(); 

    const token = generateToken(firstUser);

    const payload = {
      query: `
        query Users {
          users(first: 1) {
            edges {
              node {
                username
              }
            }
          }
        }
      `,
      variables: {}
    };

    const response = await request(createApp().callback())
      .post("/graphql")
      .set({
        Accept: "application/json",
        'Content-Type': "application/json",
        Authorization: `Bearer ${token}`
      })
      .send(JSON.stringify(payload));

    const users: any[] = response.body.data.users.edges;

    expect(response.status).toBe(200);
    expect(users.length).toBe(1)
    expect(users[0].node.username).toBe("user#1");
  });

  it(
    "should return unauthorized error when token is not provided",
    async () => {
      const payload = {
        query: `
          query User {
            users { __typename }
          }
        `,
        variables: {}
      };

      const response = await request(createApp().callback())
        .post("/graphql")
        .set({
          Accept: "application/json",
          'Content-Type': "application/json",
        })
        .send(JSON.stringify(payload));

      expect(response.status).toBe(401);
      expect(response.body.errors[0].message).toBe("User not authorized")
    });
  }
);
