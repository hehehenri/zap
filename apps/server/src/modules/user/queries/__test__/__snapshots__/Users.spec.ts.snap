// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user/queries/users should return a collection of all users 1`] = `
{
  "users": {
    "edges": [
      {
        "node": {
          "username": "user#1",
        },
      },
    ],
  },
}
`;

exports[`user/queries/users should return unauthorized error when token is not provided 1`] = `
[
  {
    "locations": [
      {
        "column": 7,
        "line": 3,
      },
    ],
    "message": "User not authorized",
    "path": [
      "users",
    ],
  },
]
`;
